contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,421 @@
+import "./IApplication.sol";
+import "../token/T.sol";
+import "../vending/VendingMachine.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+/// @notice Meta staking contract. 3 roles: app, meta staking, app manager
+contract TokenStaking is Ownable {
+    using SafeERC20 for T;
+
+    // TODO events
+
+    enum StakingProvider {
+        NU,
+        KEEP,
+        T
+    }
+
+    struct AppAllocation {
+        uint256 allocated;
+        uint256 endDeallocation;
+        uint256 deallocated;
+    }
+
+    struct ApplicationInfo {
+        mapping(StakingProvider => bool) availability;
+        uint256 allocatedOverall;
+        uint256 deallocationDuration;
+        uint256 minAllocationSize;
+    }
+
+    T public immutable token;
+    IKeepTokenStaking public immutable keepStakingContract;
+    INuCypherStakingEscrow public immutable nucypherStakingContract;
+    VendingMachine public immutable keepVendingMachine;
+    VendingMachine public immutable nucypherVendingMachine;
+
+    mapping(address => uint256) public tStakers;
+
+    mapping(address => mapping(StakingProvider => mapping(address => AppAllocation)))
+        public allocationsPerStaker;
+    // TODO public getter for availability
+    mapping(address => ApplicationInfo) public appInfo;
+    address[] public apps;
+
+    /// @param _token Address of T token contract
+    /// @param _keepStakingContract Address of Keep staking contract
+    /// @param _nucypherStakingContract Address of NuCypher staking contract
+    /// @param _keepVendingMachine Address of Keep vending machine
+    /// @param _nucypherVendingMachine Address of NuCypher vending machine
+    constructor(
+        T _token,
+        IKeepTokenStaking _keepStakingContract,
+        INuCypherStakingEscrow _nucypherStakingContract,
+        VendingMachine _keepVendingMachine,
+        VendingMachine _nucypherVendingMachine
+    ) {
+        // TODO check contracts and input variables
+        token = _token;
+        keepStakingContract = _keepStakingContract;
+        nucypherStakingContract = _nucypherStakingContract;
+        keepVendingMachine = _keepVendingMachine;
+        nucypherVendingMachine = _nucypherVendingMachine;
+    }
+
+    /// @notice Penalizes staker `staker`; the penalty details are encoded in `penaltyData`
+    function slashStaker(address staker, bytes calldata penaltyData) external {
+        // TODO check slashing event (from app)
+        // TODO choose stake (and check allocation)
+        // TODO prepare caldata
+        // TODO send slashing call
+    }
+
+    /// @notice Enable/disable application for the specified staking providers and reset deallocation duration
+    function setupApplication(
+        address app,
+        StakingProvider[] calldata stakingProviders,
+        bool availability
+    ) external onlyOwner {
```

Comment:
I think we should consider `setupApplication` to always provide application address not known before (so it is always a setup from scratch) and - optionally - we should have a separate function to enable availability for the application for the particular staking provider.

I think we should not allow to disable the availability as it is leading to other problems, like - what is happening with existing stakers - can they still perform the work? That's also why I marked additional function for enabling the availability as optional. If it is an operation that can not be reversed, we should assume that the governance call to `setupApplication` is well-thought-out.

Reply:
✔️